Address Manager

ConnorBryan opened the pull request:
### Description
Adds functionality related to an "Address Book" to MyCrypto:
* An "Address Book" tab after logging in where new <Address: Label> pairs can be entered.
![address-book-label](https://user-images.githubusercontent.com/10479826/39088543-dbba999c-4568-11e8-88b0-05fa6ec8aeb0.png)

* The `AccountInfo` component on the right side of the account now displays one of three states based on whether the user has entered a label for their own address:

1. No label, a button shows prompting the quick creation.

![add-label-account-info](https://user-images.githubusercontent.com/10479826/39088562-1d0a6bde-4569-11e8-837c-892322c321e2.png)

2. A label exists, shows an edit icon as well as the label.

![screen shot 2018-04-21 at 1 37 42 pm](https://user-images.githubusercontent.com/10479826/39088565-2f268faa-4569-11e8-9cbc-5e28db4ada44.png)

3. The label is being edited from `AccountInfo`

![screen shot 2018-04-21 at 1 38 10 pm](https://user-images.githubusercontent.com/10479826/39088567-3f56660c-4569-11e8-9458-8cd98463fa16.png)

* The `AddressField` component now displays a dropdown when the value contains two or more characters. This dropdown is navigable via arrow keys and will enter the address into the field on selection.

![screen shot 2018-04-21 at 1 40 10 pm](https://user-images.githubusercontent.com/10479826/39088593-ac0fb4a6-4569-11e8-8edf-af3250932f84.png)

* If the value in the `AddressField` component matches a label, the label for the input field will change.

![screen shot 2018-04-21 at 1 42 09 pm](https://user-images.githubusercontent.com/10479826/39088603-e81a7ae4-4569-11e8-8845-3ff22f3ecfd3.png)


* Addresses in the `DeterministicWalletModal` now display a label on top if one exists.

![deterministic-wallet-modal](https://user-images.githubusercontent.com/10479826/39088534-c125748a-4568-11e8-84f1-c316941f9b9a.png)

### Changes
* Created a new tab in `Tabs/SendTransaction`: `AddressBook`
* Created new components: AddressBookTable, AddressBookTableRow, AddressLabel
* Created an addition to `AddressFieldFactory` that acts as a dropdown
* Implemented a fully tested `addressBook` reducer that adds labels to both LocalStorage and the redux state
* Added new tested selectors to get pieces of `addressBook`

### Steps to Test
1. Log into an account
2. Navigate to the `Address Book` tab
3. Click the "+" button next to the Add new address/label row
* The `Add new address` field should be selected
4. Enter an invalid address
5. Click the "+" button again
* The `Add new address` field should be selected again
6. Enter a valid Ethereum address
7. Enter the label "Minni"
8. Press <ENTER>
* A new row should be added to the table with the entered Address/Minni pair.
9. Click on the Minni field, change it to Minnie, press <ENTER>
* The field should no longer be highlighted.
10. On the `AccountInfo` component, press the `Add Label` button.
11. In the input, enter `Mine` and press enter.
12. The label should now show `Mine`
13. The `AddressBook` table in the active tab should now have a new entry for `Mine`.
14. Press the `X` button next to `Minnie` label.
* It should be removed.
15. Navigate to the `Send Ether & Tokens` tab.
16. Enter `Mi` into the To field.
* A dropdown should appear showing `Mine` with the address.
17. Press the <DOWN ARROW> key, then the <ENTER> key.
18. The address for `Mine` should populate the field and the label should read (check) Mine

HenryNguyen5 commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,178 @@
+  private handleSave = (addressToLabel: AddressLabelPair) => {
+    const { label, address } = addressToLabel;
+
+    this.props.addLabelForAddress({ label, address });
```
No need to destructure here, just pass the arg along

ConnorBryan replied:
👍

HenryNguyen5 commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,178 @@
+
+  private setEditingRow = (editingRow: number | null) => this.setState({ editingRow });
+
+  private removeEntry = (address: string) => this.props.removeLabelForAddress(address);
```
The two above methods dont really seem needed, can just call the associated functionality directly

ConnorBryan replied:
I'll agree on `removeEntry`, but I like keeping `setState` calls in their own method as a personal quirk. I don't think it causes harm, and it improves readability.

:+1: on `removeEntry`

HenryNguyen5 commented on common/selectors/transactions.ts:
```diff
@@ -28,3 +28,7 @@ export function getRecentWalletTransactions(state: AppState): SavedTransaction[]
   }
 }
+
+export function getTransactionToRaw(state: AppState) {
```
Almost duplicate of `getTo` in `common/selectors/transaction/fields`, you can just extend `getTo` to return `raw` in another function

ConnorBryan replied:
:+1:

HenryNguyen5 commented on common/selectors/transactions.ts:
```diff
@@ -28,3 +28,7 @@ export function getRecentWalletTransactions(state: AppState): SavedTransaction[]
 }
+
+export function getTransactionToRaw(state: AppState) {
+  return state.transaction.fields.to.raw;
```
Also the selectors in this file are for `state.transactions` not `state.transaction`

ConnorBryan replied:
:+1:

HenryNguyen5 commented on common/components/BalanceSidebar/AccountInfo.tsx:
```diff
@@ -156,16 +160,16 @@ class AccountInfo extends React.Component<Props, State> {
               >
                 <UnitDisplay
-                  value={balance.wei}
+                  value={etherBalance}
```
Unrelated code refactor while I was cleaning up the types in this component. My bad.

ConnorBryan replied:
👻

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldDropdown.scss:
```diff
@@ -0,0 +1,42 @@
+
+.AddressFieldDropdown {
+  position: absolute;
+  top: calc(100% - 16px);
```
This should also be the width of the input, so it should be a in a position: relative div around the input, and have `left: 0; width: 100%;`

ConnorBryan replied:
Top comment: :+1: , though I just used `$space` since the margin-bottom is set in `input-group-input`.

Bottom comment: 👍

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldDropdown.scss:
```diff
@@ -0,0 +1,42 @@
+      background: $gray-lightest;
+    }
+
+    strong:first-of-type {
```
Prefer classnames.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldDropdown.scss:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    strong:first-of-type {
+      width: $space-xl * 6;
```
Rather than give labels a fixed width, which can look weird if they're all short or you only have one, I'd suggest doing

(Identicon) **Label** _address_

So that it can just be left aligned as normal.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldDropdown.tsx:
```diff
@@ -0,0 +1,160 @@
+          return this.handleDownArrowKeyDown();
+        default:
+          return;
+      }
```
This is dope, make sure you also preventDefault() if it's up / down / enter.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressField.tsx:
```diff
@@ -14,23 +14,37 @@ interface Props {
+          <span>
+            <i className="fa fa-check" /> {currentLabel}
+          </span>
+        );
```
I'm not wild about changing the label for accessibility reasons. I'd prefer this be some green text below the input or something. Ideally this can be unified with the ENS resolver, and share the same styles. That one could use some UI love.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldFactory.tsx:
```diff
@@ -33,16 +48,36 @@ class AddressFieldFactoryClass extends React.Component<Props> {
+    const { isFocused } = this.state;
+
+    // Allow the click event on a dropdown item to process before hiding the dropdown.
+    const onBlur = () => (this.goingToBlur = window.setTimeout(this.blur, 150));
```
This should be a class function, otherwise it'll always cause a re-render since it's a new function each time.

ConnorBryan replied:
:+1:

wbobeirne commented on shared/keycodes.ts:
```diff
@@ -0,0 +1,11 @@
+  DOWN_ARROW: 40
+};
+
+export default KeyCodes;
```
React provides some more readable keyboard event attributes that don't require the mapping https://reactjs.org/docs/events.html#keyboard-events

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  private handleSave = (addressToLabel: AddressLabelPair) => {
+    const { label, address } = addressToLabel;
```
Likewise, we should not allow for duplicate labels. Perhaps this should be handled as a saga.

dternyak commented:
@wbobeirne I like the touchups quite a bit! Would be great to get those changes in.

ConnorBryan commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,178 @@
+                  onChange={this.setTemporaryAddress}
+                  setInnerRef={this.setAddressInputRef}
+                />
+                <Identicon address={temporaryAddress} />
```
:+1:

ConnorBryan commented on common/components/AddressBookTableRow.tsx:
```diff
@@ -0,0 +1,90 @@
+              className="btn btn-sm  btn-danger"
+              onClick={onRemoveClick}
+            >
+              <i className="fa fa-close" />
```
:+1:

ConnorBryan commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,178 @@
+            <td>
+              <div className="AddressBookTable-cell">
+                <button className="btn btn-sm btn-default" onClick={this.handleAddEntry}>
+                  <i className="fa fa-plus" />
```
:+1:

james-prado commented on common/translations/lang/en.json:
```diff
@@ -1,6 +1,15 @@
   "data": {
+    "ADDRESS": "Address",
+    "LABEL": "Label",
+    "INVALID_LABEL_LENGTH": "An address label must be between 2 and 20 characters.",
```
Why are we imposing a character limit for labels? If responsiveness is an issue, we can always truncate the labels.

james-prado replied:
Also, addresses should be truncated on smaller screens. Personally I like the way metamask does it, they truncate the middle section of an address so you can verify the first 3 and last 3 digits of an address.

ConnorBryan replied:
I imposed the limit because I was asked to by will. Truncation is fine, but again, in the interest of avoiding story bloat that more advanced functionality might be better as a different story.

ConnorBryan replied:
Fixed: Updated to 50 and introduced truncation in the various spots that display the label.

james-prado commented on common/components/AddressFieldFactory/AddressFieldDropdown.scss:
```diff
@@ -0,0 +1,42 @@
+  z-index: 1;
+  list-style-type: none;
+  margin: 0;
+  padding: 0;
```
I'm actually the opposite, I think that dropdown style throughout the site should remain consistent. Maybe this would be a good time to change how we do dropdowns / input styles?

All the other dropdowns just follow the simple input style, but I'm definitely open to updating those. I'm also not opposed to the drop-shadow, but I do think dropdowns should match the input style.

ConnorBryan replied:
That may be the case, but the scope on this has already exploded. This would probably be better as a future story.

james-prado commented on common/components/AddressFieldFactory/AddressFieldDropdown.tsx:
```diff
@@ -0,0 +1,157 @@
+          role="option"
+          title={`${translateRaw('SEND_TO')}${label}`}
+        >
+          <Identicon address={address} size="2rem" />
```
On smaller screens the EthBlockies icon still gets squished 
<img width="574" alt="screen shot 2018-05-01 at 3 31 38 pm" src="https://user-images.githubusercontent.com/13836920/39489835-d0eec170-4d54-11e8-8f82-04ea941591ed.png">

ConnorBryan replied:
Whoops — I need to wrap those in containers with a minimum width. :+1:

ConnorBryan replied:
Fixed 👍 

<img width="331" alt="screen shot 2018-05-02 at 3 13 41 pm" src="https://user-images.githubusercontent.com/10479826/39546867-6de2bfec-4e1b-11e8-9b33-75181fd1b731.png">

james-prado commented on common/components/AddressFieldFactory/AddressFieldFactory.tsx:
```diff
@@ -15,16 +16,28 @@ interface OwnProps {
   readOnly: boolean;
   currentTo: ICurrentTo;
   onChange(ev: React.FormEvent<HTMLInputElement>): void;
+  onFocus(ev: React.FormEvent<HTMLInputElement>): void;
```
The onFocus for the address input should be updated so it isn't `invalid` when searching for an address in your address book. 
<img width="912" alt="screen shot 2018-05-01 at 3 33 27 pm" src="https://user-images.githubusercontent.com/13836920/39489957-305a44ae-4d55-11e8-8174-131852a7f970.png">


This occurs after:
- Begin searching for a valid address
- leave the input (`onBlur()`)
- focus the input (`onFocus()`)

james-prado replied:
Also, if you start typing an address label that doesn't exist, it should display an error, something like 'That address doesn't exist in your address book'.

ConnorBryan replied:
Agree on the bottom, but I was told that an I out should show invalid if it’s invalid. Even if you’re searching for a label, at that moment it’s not actually valid.

Clarification, @dternyak ?

dternyak replied:
@ConnorBryan The input that I mentioned that should have invalid styling when invalid is the address book address input, not the To Address input.

ConnorBryan replied:
@dternyak

Okay! I’ll get this fixed up when I get home from the hospital later.

james-prado commented on common/components/AddressBookTable.scss:
```diff
@@ -0,0 +1,85 @@
+@import 'common/sass/variables/spacing';
+@import 'common/sass/variables/colors';
+
+.AddressBookTable {
```
The Address book is still lacking some support for smaller screens.

<img width="677" alt="screen shot 2018-05-01 at 3 42 22 pm" src="https://user-images.githubusercontent.com/13836920/39490354-9cfe32b8-4d56-11e8-8b9c-bbd5416b2631.png">

ConnorBryan replied:
Oh, that must be right before it collapses. Should be a quick fix.

ConnorBryan commented:
@dternyak

I like the idea of incorporating validation to input factory in a future story. For now I can spend some time making the notification that displays in submit appear next to the label, aligned right, colored $beand-danger. Would that be acceptable?

dternyak commented:
Agreed, the input/error factory should be part of a future story.

r/e alignment: I would checkout what we do on the swap view when errors are exceed max or are less than the minimum. I think this is reasonable, although I'm sure improvements can be had all around.

<img width="726" alt="screen shot 2018-05-01 at 3 29 20 pm" src="https://user-images.githubusercontent.com/7861465/39492197-721d5986-4d54-11e8-9e72-9b7cdb814941.png">

james-prado commented:
@dternyak Yea, I think with all the updates since the last time all the inputs got some attention it might be time to clean them up again. I'll open up a new issue for updating and standardizing all our input types like inline dropdown, searchable dropdown, inline loading msgs, etc... Also standardizing error msgs and how they're displayed will be covered there too.

ConnorBryan commented:
# AddressFieldDropdown

## AddressFieldDropdown, desktop, match exists
<img width="845" alt="screen shot 2018-05-03 at 11 24 22 pm" src="https://user-images.githubusercontent.com/10479826/39612920-2a1f3a4c-4f29-11e8-89ff-0dc9c0c5eae3.png">

_Note: The dropdown can be interacted with via the up and down arrow keys, as well as `Enter` for selection._

## AddressFieldDropdown, desktop, no match exists
<img width="837" alt="screen shot 2018-05-03 at 11 27 13 pm" src="https://user-images.githubusercontent.com/10479826/39612959-88634562-4f29-11e8-874d-0b2eba8dc808.png">

## AddressFieldDropdown, desktop, invalid address entry
<img width="850" alt="screen shot 2018-05-03 at 11 28 50 pm" src="https://user-images.githubusercontent.com/10479826/39613000-cc86a036-4f29-11e8-8ecc-2c7136a3014d.png">

_Note: Entering any text other than '0x' as the first characters of the query will disable validation highlighting, as it is assumed the user is searching for a label._

## AddressFieldDropdown, mobile, match exists
<img width="174" alt="screen shot 2018-05-03 at 11 25 43 pm" src="https://user-images.githubusercontent.com/10479826/39612937-5c3ce970-4f29-11e8-9200-9077d2c78af2.png">

## AddressFieldDropdown, mobile, no match exists
<img width="191" alt="screen shot 2018-05-03 at 11 27 54 pm" src="https://user-images.githubusercontent.com/10479826/39612977-a1b49d22-4f29-11e8-9d60-6685d900bb30.png">


# AddressBookTable

## AddressBookTable, desktop, default view
<img width="866" alt="screen shot 2018-05-03 at 11 48 52 pm" src="https://user-images.githubusercontent.com/10479826/39613385-98ed3c50-4f2c-11e8-9e0a-4dfea1d7b9c5.png">

## AddressBookTable, desktop, with errors
<img width="862" alt="screen shot 2018-05-03 at 11 49 50 pm" src="https://user-images.githubusercontent.com/10479826/39613399-b8deb7a0-4f2c-11e8-80f9-813357b2b970.png">

## AddressBookTable, desktop, submitting with errors
<img width="929" alt="screen shot 2018-05-03 at 11 50 33 pm" src="https://user-images.githubusercontent.com/10479826/39613410-c96d9604-4f2c-11e8-877e-316bd857b078.png">

## AddressBookTable, desktop, error on row
<img width="860" alt="screen shot 2018-05-03 at 11 51 11 pm" src="https://user-images.githubusercontent.com/10479826/39613464-f357e262-4f2c-11e8-8fd1-4cfa78587959.png">

## AddressBookTable, mobile, default view
<img width="190" alt="screen shot 2018-05-03 at 11 52 34 pm" src="https://user-images.githubusercontent.com/10479826/39613482-1815c92a-4f2d-11e8-81c9-8b994a3bc9b8.png">

## AddressBookTable, mobile, with errors
<img width="183" alt="screen shot 2018-05-03 at 11 53 17 pm" src="https://user-images.githubusercontent.com/10479826/39613495-2ef0c1e0-4f2d-11e8-9744-b89104136d04.png">

## AddressBookTable, mobile, submitting with errors
<img width="194" alt="screen shot 2018-05-03 at 11 53 52 pm" src="https://user-images.githubusercontent.com/10479826/39613508-3fbb16a6-4f2d-11e8-9d15-a064167c97eb.png">

## AddressBookTable, mobile, error on row
<img width="187" alt="screen shot 2018-05-03 at 11 54 46 pm" src="https://user-images.githubusercontent.com/10479826/39613520-6069deaa-4f2d-11e8-9fc9-04e1873d3e7b.png">


# AccountAddress

## AccountAddress, default view
<img width="407" alt="screen shot 2018-05-03 at 11 55 24 pm" src="https://user-images.githubusercontent.com/10479826/39613533-7fea0aca-4f2d-11e8-8126-151a66cfbc4d.png">

## AccountAddress, editing view
<img width="411" alt="screen shot 2018-05-03 at 11 56 24 pm" src="https://user-images.githubusercontent.com/10479826/39613552-9f341984-4f2d-11e8-9ba9-e1c795bbfcf5.png">

## AccountAddress, no label
<img width="411" alt="screen shot 2018-05-03 at 11 56 55 pm" src="https://user-images.githubusercontent.com/10479826/39613559-b2c02f2e-4f2d-11e8-8529-f04192d86fce.png">

## AccountAddress, error
<img width="408" alt="screen shot 2018-05-03 at 11 57 24 pm" src="https://user-images.githubusercontent.com/10479826/39613571-c6a541a0-4f2d-11e8-8ff1-ebad5e2a1b87.png">

## AccountAddress, submitting with error
<img width="424" alt="screen shot 2018-05-03 at 11 58 03 pm" src="https://user-images.githubusercontent.com/10479826/39613584-dabdc8ec-4f2d-11e8-842a-a1bfb3700c72.png">

# Notes

Okay, I'm pretty sure I've addressed all comments. I've attached updated pictures of the various state. Please let me know any other changes you'd like made prior to merge.

@dternyak 
@HenryNguyen5 
@wbobeirne 
@james-prado

james-prado commented:
@ConnorBryan The reason for changing the validation so it only occurs after the first time an input has blurred was to reduce the amount of error messages being displayed when a user hadn't even had a change to fill in the content for a given input. Ideally, the user should have a chance to fill in the content correctly before throwing errors at them. The only exception to this is if a user tries to submit a form without filling all the required inputs.

wbobeirne commented on common/components/AddressBookTable.tsx:
```diff
@@ -0,0 +1,316 @@
+    const labelInputClassName = labelTouchedWithError ? 'invalid' : '';
+    const nonMobileTemporaryInputErrorClassName =
+      'AddressBookTable-row-error-temporary-input--non-mobile';
+    const nonMobileVisibleErrorClassName = `${nonMobileTemporaryInputErrorClassName}--visible`;
```
For quick little modifiers like these, we have a precedent of using state classes like `is-visible` or `is-non-mobile` to save some line length.

ConnorBryan replied:
@wbobeirne Can you point to an example?

wbobeirne replied:
https://github.com/MyCryptoHQ/MyCrypto/blob/aa0f9cd455aab2bd7134036fee858de4ae10028c/common/components/NavigationLink.tsx#L28 is an example, here's the SuitCSS guide on the matter https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md#is-stateOfComponent

wbobeirne commented on common/components/AddressBookTableRow.tsx:
```diff
@@ -0,0 +1,149 @@
+              </label>
+              <Input
+                name={labelName}
+                title={`${translateRaw('EDIT_LABEL_FOR')}${address}`}
```
There should be a space between these two.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressBookTableRow.tsx:
```diff
@@ -0,0 +1,149 @@
+        </div>
+        {labelError && (
+          <div className="AddressBookTable-row AddressBookTable-row-error AddressBookTable-row-error--non-mobile">
+            <div />
```
What're these empty divs for? If you're trying to make something display: block, better to do that in CSS.

ConnorBryan replied:
:+1:

wbobeirne commented on common/components/AddressFieldFactory/AddressFieldDropdown.tsx:
```diff
@@ -0,0 +1,165 @@
+    const { currentTo } = this.props;
+    const noMatchContent = currentTo.startsWith('0x') ? null : (
+      <li className="AddressFieldDropdown-dropdown-item AddressFieldDropdown-dropdown-item-no-match">
+        <i className="fa fa-warning" /> {translate('NO_LABEL_FOUND_CONTAINING')}"{currentTo}".
```
Should be a space between translate and quotes.

wbobeirne replied:
For context, I think that some of the translations have trailing spaces, but I don't want to rely on that in case someone forgets to add one when doing the translations in other languages. Better to do it once for all cases.

ConnorBryan replied:
:+1:

avanimiriyala commented:
Hey Team, I have one small recommendation here in the design. In the Address Book, is it possible to move the wallet avatars to the left instead of keeping them in the middle? Not a huge UX issue, but more of a design polish thing. See attached mockup I made :) 
<img width="1082" alt="address-book" src="https://user-images.githubusercontent.com/40361706/41718756-92277ae4-7523-11e8-91e7-c4b1b1e36bb5.png">
